server/src/main/java/org/elasticsearch/indices/ExecutorSelectorService.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices;
+
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Objects;
+
+/**
+ * Some operations need to use different executors for different index patterns.
+ * Specifically, some operations on system indices are considered critical and
+ * should use the "system_critical_read" or "system_critical_write" thread pools
+ * rather than the "system_read" or "system_write" thread pools.
+ */
+public class ExecutorSelectorService {
+
+    private final SystemIndices systemIndices;
+
+    public ExecutorSelectorService(SystemIndices systemIndices) {
+        this.systemIndices = systemIndices;
+    }
+
+    /**
+     * The "get" executor should be used when retrieving documents by ID.
+     * @param indexName Name of the index
+     * @return Name of the executor to use for a get operation.
+     */
+    public String getGetExecutor(String indexName) {
```

Comment:
the `getGet` naming is kind of hard to think about. Maybe we could drop the `get` and just have methods like `executorForGet`, `executorForSearch`, `executorForWrite`?